src/Microsoft.ApplicationInsights/DataContracts/ExceptionInfo.cs

```diff
@@ -0,0 +1,85 @@
+﻿namespace Microsoft.ApplicationInsights.DataContracts
+{
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Extensibility.Implementation;
+    using Extensibility.Implementation.External;
+
+    /// <summary>
+    /// Wrapper class for <see cref="ExceptionData"/> that lets user provide exception data without having the actual Exception object.
+    /// </summary>
+    internal sealed class ExceptionInfo
+    {
+        private readonly ExceptionData data;
+        
+        /// <summary>
+        /// Constructs the instance of <see cref="ExceptionInfo"/>
+        /// </summary>
+        public ExceptionInfo(IEnumerable<ExceptionDetailsInfo> exceptionDetailsInfoList, SeverityLevel? severityLevel, string problemId,
+            IDictionary<string, string> properties, IDictionary<string, double> measurements)
+        {
+            this.data = new ExceptionData
+            {
+                exceptions = exceptionDetailsInfoList.Select(edi => edi.ExceptionDetails).ToList(),
+                severityLevel = severityLevel.TranslateSeverityLevel(),
+                problemId = problemId,
+                properties = new ConcurrentDictionary<string, string>(properties),
+                measurements = new ConcurrentDictionary<string, double>(measurements)
+            };
+        }
+
+        internal ExceptionInfo(ExceptionData data)
+        {
+            this.data = data;
+        }
+
+        /// <summary>
+        /// Gets a list of <see cref="ExceptionDetailsInfo"/> to modify as needed.
+        /// </summary>
+        public IReadOnlyList<ExceptionDetailsInfo> ExceptionDetailsInfoList => this.data.exceptions.Select(ed => new ExceptionDetailsInfo(ed)).ToList().AsReadOnly();
```

Comment:
why do we need all these allocations on property get? Can we return raw list?

Reply:
The actual details that are being serialized are stored in a collection of ExceptionDetails, which is an internal type we don't want to expose here. So we have to wrap each element in ExceptionDetailsInfo (the external counterpart) before exposing. Hence the allocations.